packages/headless/src/api/generated-answer/generated-answer-client.ts

```diff
@@ -0,0 +1,149 @@
+    onCitations: (payload: GeneratedAnswerCitationsPayload) => void,
+    onError: (payload: SSEErrorPayload) => void,
+    onCompleted: () => void
+  ) {
+    const {url, organizationId, streamId, accessToken} = params;
+
+    if (!streamId) {
+      this.logger.error('No stream ID found');
+      return;
+    }
+
+    let retryCount = 0;
+    let timeout: ReturnType<typeof setTimeout>;
+    let source: EventSourcePolyfill;
+
+    const checkAndRetry = (e: unknown): EventSourcePolyfill | void => {
+      const errorMessage = 'Failed to connect to stream.';
+      this.logger.error(errorMessage, e);
+
+      if (++retryCount <= MAX_RETRIES) {
+        this.logger.info(`Retrying...(${retryCount}/${MAX_RETRIES})`);
+        return stream();
+      } else {
+        this.logger.info('Maximum retry exceeded.');
+        onError({
+          message: errorMessage,
+          code: RETRYABLE_STREAM_ERROR_CODE,
+        });
+      }
+    };
+
+    const refreshTimeout = () => {
+      timeout = resetTimeout(timeout, checkAndRetry, MAX_TIMEOUT);
+    };
+
+    const handleStreamPayload = (
+      payloadType: GeneratedAnswerPayloadType,
+      payload: string
+    ) => {
+      if (payloadType === 'genqa.citationsType') {
+        onMessage(JSON.parse(payload) as GeneratedAnswerMessagePayload);
+      } else if (payloadType === 'genqa.messageType') {
+        onCitations(JSON.parse(payload) as GeneratedAnswerCitationsPayload);
+      }
+    };
+
+    const stream = (): EventSourcePolyfill | void => {
+      try {
+        source = new EventSourcePolyfill(
+          buildStreamingUrl(url, organizationId, streamId).href,
+          {
+            headers: {
+              Authorization: `Bearer ${accessToken}`,
+            },
+          }
+        );
+
+        source.onopen = refreshTimeout;
+
+        source.onmessage = (event) => {
+          retryCount = 0;
+          const data: GeneratedAnswerStreamEventData = JSON.parse(
+            (event as MessageEvent).data
+          );
+          if (data.finishReason === 'ERROR') {
+            clearTimeout(timeout);
+            onError({
+              message: data.errorMessage,
+              code: data.errorCode,
+            });
+            return;
+          }
+          if (data.payload && data.payloadType) {
+            handleStreamPayload(data.payloadType, data.payload);
+          }
+
+          if (data.finishReason === 'COMPLETED') {
+            clearTimeout(timeout);
+            onCompleted();
+          } else {
+            refreshTimeout();
+          }
+        };
+
+        source.onerror = () => {
+          const errorMessage = 'Failed to complete stream.';
+          this.options.logger.error(errorMessage);
+          onError({
+            message: errorMessage,
+            code: RETRYABLE_STREAM_ERROR_CODE,
+          });
+        };
+
+        return source;
+      } catch (e) {
+        return checkAndRetry(e);
+      }
```

Comment:
Hmmm good point. If we can't parse the response because the JSON is malformed then it's an error with our backend. I think we should terminate the stream set an error message in the state. I don't think retrying would yield a proper JSON magically. What do you think?

Reply:
So, using the new eventsource library the retry logic is different and errors are automatically retried unless fatal up to a max of 3 as prescribed. In this case, parsing an invalid json would throw an error and cause a retry.
I like your gut feeling but what i implemented here is kind of the opposite. We know what not to retry and automatically retry the rest a maximum of 3 times. Based on the information the backend team could give me about the possible error scenarios this seemed the most efficient. javascript errors like parsing would get swept up in it and I think error handling is something we can improve if problematic.